Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
         return;
     }
 
-    auto* document = dynamicDowncast<Document>(context);
-    if (!document) {
-        promise.reject(Exception { NotSupportedError });
-        return;
-    }
-
-    if (!document->isFullyActive()) {
+    auto* document = dynamicDowncast<Document>(*context);
+    if (document && !document->isFullyActive()) {
         promise.reject(Exception { InvalidStateError, "The document is not fully active"_s });
-        return;
-    }
-
-    if (!document->page()) {
-        promise.reject(Exception { NotSupportedError });
-        return;
-    }
-
-    if (!permissionDescriptorValue) {
-        promise.reject(Exception { DataError, "The parameter is invalid"_s });
-        return;
+        return; 
     }
 
     JSC::VM& vm = context->globalObject()->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
+    auto permissionDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
     if (UNLIKELY(scope.exception())) {
         promise.reject(Exception { ExistingExceptionError });
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
+    auto* origin = context->securityOrigin();
+    auto originData = origin ? origin->data() : SecurityOriginData { };
+
+    if (document) {
+        if (!document->page()) {
+            promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
+            return;
+        }
+
+        if (!isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
+            promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
+            return;
+        }
+
+        PermissionController::shared().query(ClientOrigin { document->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [document = Ref { *document }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+            if (!permissionState)
+                promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
+            else
+                promise.resolve(PermissionStatus::create(document, *permissionState, permissionDescriptor));
         });
         return;
     }
 
-    auto* origin = context->securityOrigin();
-    auto originData = origin ? origin->data() : SecurityOriginData { };
+    auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context);
+    ASSERT(workerGlobalScope);
```

Comment:
Then you can drop this assertion since downcast<> will assert for you.

Reply:
Fixed!